scipy/interpolate/_rbfinterp.py

```diff
@@ -0,0 +1,538 @@
+"""Module for RBF interpolation"""
+import warnings
+from functools import lru_cache
+from itertools import combinations_with_replacement
+
+import numpy as np
+from numpy.linalg import LinAlgError
+from scipy.spatial import KDTree
+from scipy.special import binom
+from scipy.linalg.lapack import dgesv
+
+from ._rbfinterp_pythran import _build_system, _evaluate, _polynomial_matrix
+
+
+__all__ = ['RBFInterpolator', 'KNearestRBFInterpolator']
+
+
+# The RBFs that are implemented
+_AVAILABLE = {'linear', 'tps', 'cubic', 'quintic', 'mq', 'imq', 'iq', 'ga'}
+
+
+# The shape parameter does not need to be specified when using these RBFs
+_SCALE_INVARIANT = {'linear', 'tps', 'cubic', 'quintic'}
+
+
+# For RBFs that are conditionally positive definite of order m, the interpolant
+# should include polynomial terms with degree >= m - 1. Define the minimum
+# degrees here. These values are from Chapter 8 of Fasshauer's "Meshfree
+# Approximation Methods with MATLAB". The RBFs that are not in this dictionary
+# are positive definite and do not need polynomial terms
+_NAME_TO_MIN_DEGREE = {
+    'mq': 0,
+    'linear': 0,
+    'tps': 1,
+    'cubic': 1,
+    'quintic': 2
+    }
+
+
+@lru_cache(maxsize=8)
+def _monomial_powers(ndim, degree):
+    """
+    Returns the powers for each monomial in a polynomial with the specified
+    number of dimensions and degree
+    """
+    out = []
+    for deg in range(degree + 1):
+        for itm in combinations_with_replacement(np.eye(ndim, dtype=int), deg):
+            out.append(sum(itm, np.zeros(ndim, dtype=int)))
+
+    if not out:
+        out = np.zeros((0, ndim), dtype=int)
+    else:
+        out = np.array(out)
+
+    return out
+
+
+def _build_and_solve_system(y, d, smoothing, kernel, epsilon, powers):
+    """
+    Build and solve the RBF interpolation system of equations
+
+    Parameters
+    ----------
+    y : (P, N) float ndarray
+        Data point coordinates
+    d : (P, S) float ndarray
+        Data values at `y`
+    smoothing : (P,) float ndarray
+        Smoothing parameter for each data point
+    kernel : str
+        Name of the RBF
+    epsilon : float
+        Shape parameter
+    powers : (R, N) int ndarray
+        The exponents for each monomial in the polynomial
+
+    Returns
+    -------
+    coeffs : (P + R, S) float ndarray
+    shift : (N,) float ndarray
+        Domain shift used to create the polynomial matrix
+    scale : (N,) float ndarray
+        Domain scaling used to create the polynomial matrix
+
```

Comment:
The descriptions are sentences and should have a point. Also `coeffs` needs the extra space after it if you don't want to add a description.

```suggestion
    Parameters
    ----------
    y : (P, N) float ndarray
        Data point coordinates.
    d : (P, S) float ndarray
        Data values at `y`.
    smoothing : (P,) float ndarray
        Smoothing parameter for each data point.
    kernel : str
        Name of the RBF.
    epsilon : float
        Shape parameter.
    powers : (R, N) int ndarray
        The exponents for each monomial in the polynomial.

    Returns
    -------
    coeffs : (P + R, S) float ndarray

    shift : (N,) float ndarray
        Domain shift used to create the polynomial matrix.
    scale : (N,) float ndarray
        Domain scaling used to create the polynomial matrix.

```